### DISCUSSION TITLE
Internet support via USB?

### DISCUSSION THREAD

**Discussion by jamiefaye** [SynthstromAudible/DelugeFirmware#364] [Ideas]
It is possible to connect a Deluge to the Internet via the USB interface in "host mode".

It would cost some memory on the Deluge, both for the appropriate network stack and for the code to allow you to manage a WiFi or Ethernet connection.

An alternative is to "tunnel via SysEx", however SysEx over the USB is considerably slower than WiFi or Ethernet. (We would probably implement a custom API for Deluge control rather than shipping IP packets back and forth).

**Comment by manysounds** [comment]
RTP-MIDI?

**Comment by PaulFreund** [comment]
I think it is a cool idea but I think flash memory will be our limiting factor because network and IP stack will probably bloat our firmware considerably. In our current hardware environment it makes more sense for the Deluge to be the slave in the interaction

**Reply by sapphire-arches** [reply]
I looked around a bit and found https://github.com/tass-belgium/picotcp/ a (claimed) small TCP implementation. It doesn't quote binary sizes but looking around it seems to be ~40k which, while not large, is perhaps large enough to be concerning. This would be required even if the Deluge itself was acting as a the USB peripheral.

Further complicating matters is the fact that there's at least 3 popular USB ethernet standards - RNDIS (microsoft), CDC-ECM (old), and CDC-NCM (new). Different USB/Ethernet peripherals advertise themselves as compatible with different standards of these. If we make the Deluge act as a peripheral, it would be easier to implement only the CDC-NCM mode as that is well-supported by all 3 major desktop operating systems.

**Reply by heeen** [reply]
how about a companion esp32 module? it has its own IP stack and could implement a simpler serial protocol that wouldn't require much code on the deluge.
